core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -382,56 +383,33 @@ void validateIncomingState(ClusterState incomingState, ClusterState lastSeenClus
             logger.warn("received a cluster state from [{}] and not part of the cluster, should not happen", incomingState.nodes().masterNode());
             throw new IllegalStateException("received state from a node that is not part of the cluster");
         }
-        // state from another master requires more subtle checks, so we let it pass for now (it will be checked in ZenDiscovery)
-        if (currentNodes.localNodeMaster() == false) {
-            ZenDiscovery.validateStateIsFromCurrentMaster(logger, currentNodes, incomingState);
-        }
 
-        if (lastSeenClusterState != null
-                && Objects.equals(lastSeenClusterState.nodes().masterNodeId(), incomingState.nodes().masterNodeId())
-                && lastSeenClusterState.version() > incomingState.version()) {
-            logger.debug("received an older cluster state from master, rejecting (received version [{}], last version is [{}])",
-                    incomingState.version(), lastSeenClusterState.version());
-            throw new IllegalStateException("cluster state version [" + incomingState.version() + "] is old (last seen version [" + lastSeenClusterState.version() + "])");
-        }
+        ZenDiscovery.validateStateIsFromCurrentMaster(logger, currentNodes, incomingState);
     }
 
     protected void handleCommitRequest(CommitClusterStateRequest request, final TransportChannel channel) {
-        ClusterState committedClusterState;
-        synchronized (lastSeenClusterStateMutex) {
-            committedClusterState = lastSeenClusterState;
-        }
-
-        // if this message somehow comes without a previous send, we won't have a cluster state
-        String lastSeenUUID = committedClusterState == null ? null : committedClusterState.stateUUID();
-        if (request.stateUUID.equals(lastSeenUUID) == false) {
-            throw new IllegalStateException("tried to commit cluster state UUID [" + request.stateUUID + "], but last seen UUID is [" + lastSeenUUID + "]");
-        }
-
-        try {
-            listener.onNewClusterState(committedClusterState, new NewClusterStateListener.NewStateProcessed() {
-                @Override
-                public void onNewClusterStateProcessed() {
-                    try {
-                        channel.sendResponse(TransportResponse.Empty.INSTANCE);
-                    } catch (Throwable e) {
-                        logger.debug("failed to send response on cluster state processed", e);
-                        onNewClusterStateFailed(e);
-                    }
+        final ClusterState state = pendingStatesQueue.markAsCommitted(request.stateUUID, new PendingClusterStatesQueue.StateProcessedListener() {
+            @Override
+            public void onNewClusterStateProcessed() {
+                try {
+                    channel.sendResponse(TransportResponse.Empty.INSTANCE);
```

Comment:
I know this code is mimicking what the old stuff did, but can you add an explanation for why an empty response is sent?